enhancements/microshift/microshift-pluggable-core-components.md

```diff
@@ -0,0 +1,213 @@
+customers to install individual compomnents as rpm packages.
+
+
+## Motivation
+
+The principal factor that gave birth to the idea of MicroShift was the need to bring
+the OpenShift deployment model into the Edge where the power, memory and CPU budgets
+are reduced.
+
+While the default configuration makes sense to most of our customers in the Device
+Edge area, still our current CSI, CNI, and router components use 35% of the memory
+footprint and a significative CPU and power footprint. Making those components
+pluggable would allow customers who don't require advanced network features,
+persistent volumes, or ingress/routes to disable or switch those technologies
+and reduce the memory/cpu and disk/over-the-wire footprint.
+
+
+### User Stories
+
+An application developer needs the whole stack of components, creates an image
+that contains:
+1. microshift
+
+Nothing would change in this case, installing microshift, which becomnes a virtual
+package, installs the necessary dependencies:
+* microshift-core
+* microshift-networking-ovn
+* microshift-storage-topolvm
+* microshift-routes
+
+as well as any other dependencies.
+
+An application developer only needs advanced networking, but no persistent volumes,
+and no routes/ingress.  Creates an image that contains:
+1. microshift-core
+2. microshift-networking-ovn
+
+
+An application developer wants the bare minimum set of components creates an image
+with:
+1. microshift-core
+
+An application developer wanting bare minimum networking but persistent volumes
+creates an image with:
+1. microshift-core
+2. microshift-storage-topolvm
+
+### Goals
+* Allow networking, storage and router core components to be optionally
+  installed.
+
+* As we refactor and re-use the kustomize package used for loading user workloads,
+  add support for loading yamls from multiple directories, i.e.:
+  * `/usr/lib/microshift/manifests.d/user-app-1/`
+  * `/usr/lib/microshift/manifests.d/user-app-2/`
+  * `/etc/microshift/manifests.d/user-app-3/`
+  * `...`
+
+* Split the RPM packages to resemble those optional components.
+
+* Preserve previous behavior in terms of RPM package installation.
+
+### Non-Goals
+
+* Making DNS component optional.
+* Making the certificate management component optional.
+
+## Proposal
+
+To allow modularity of the described core components, the following changes
+would be introduced:
+
+Update the service manager to allow core components as kustomization yamls,
+in the same way we already support for user application manifests. i.e.:
+
+`/usr/lib/microshift/components.d/networking-ovn/kustomize.yaml`
+`/usr/lib/microshift/components.d/storage-topolvm/kustomize.yaml`
+`/usr/lib/microshift/components.d/openshift-router/kustomize.yaml`
+
+  MicroShift component manager would scan the `/usr/lib/microshift/components.d/`
+  directory and extract metadata from the yamls to analyze dependencies between
+  core components at boot. Then start the right componments in the right order.
```

Comment:
That setting would be controlled by the RPMs; the RPMs would contain the specific files which control that, i.e.:
https://github.com/openshift/microshift/blob/main/packaging/rpm/microshift.spec#L148

We could control that network sub-packages do not overlap, for example, if we add a conflicts clause between microshift-networking-ovn and any other CNI. It would be written as "Conflict: microshift-networking-bridge" if we needed to create a package for the bridge. Then the developer can install one or the other, but not both.

Reply:
Or we could put some constraints (at least for phase one in 4.13) that only one CNI can be installed (constraint documented and runtime check if there's more than 1 CNI installed? I don't see a way to prohibit installation of arbitrary RPM or even just moving the files into specific dir).

Not sure if it's still a problem, but I remember (at least upstream) multus was scanning for master CNI only on start, so it had to be deployed second.

Reply:
I'd simply it by telling microshift which plugin to use and let microshift to manipulate the crio config. it is less error prone when we have one component touching the config.

Reply:
Yes, but one of the design decissions in MicroShift was that MicroShift would not modify the system configuration (/etc) on runtime.

Using rpms it's done on image creation.

Reply:
makes sense.